feat(bytesRead-to-bytesWritten): implement Node.js DEP0108 migration

Eliekhoury17 opened the pull request:
This pull request implements the codemod to replace deprecated `zlib.bytesRead` 
with `zlib.bytesWritten` in Node.js transform streams, addressing DEP0108 (#196).

Changes included:
- Updated the codemod workflow to track zlib stream variables and function parameters.
- Added tests for assigned variables, destructured imports, and trackStreamProgress.
- Updated README.md and package.json metadata.
- Ensured all tracked usages of `.bytesRead` are replaced correctly.

This PR is part of the feature branch `feat/bytesread-to-bytesWritten`.

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+	const edits: Edit[] = [];
+	const linesToRemove: Range[] = [];
+
+	// 1️ Find all zlib imports/requires
```
```suggestion
	// 1 Find all zlib imports/requires
```

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+		}
+	}
+
+	// 2️ Track variables assigned from factories
```
```suggestion
	// 2 Track variables assigned from factories
```

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+	}
+
+
+	// 3️ Replace .bytesRead → .bytesWritten for tracked variables
```
```suggestion
	// 3 Replace .bytesRead → .bytesWritten for tracked variables
```

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+	// 3️ Replace .bytesRead → .bytesWritten for tracked variables
+	for (const variable of streamVariables) {
+		const matches = rootNode.findAll({ rule: { pattern: `${variable}.bytesRead` } });
+		for (const match of matches) {
```
```suggestion

		for (const match of matches) {
```

NIT

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+		}
+	}
+
+	// 4️ Replace .bytesRead → .bytesWritten for function parameters
```
```suggestion
	// 4 Replace .bytesRead → .bytesWritten for function parameters
```

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	if (!edits.length) return null;
+	return removeLines(rootNode.commitEdits(edits), linesToRemove);
```
```suggestion

	return removeLines(rootNode.commitEdits(edits), linesToRemove);
```

NIT

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+	// 2️ Track variables assigned from factories
+	for (const factory of factoryBindings) {
+		const matches = rootNode.findAll({ rule: { pattern: `const $$$VAR = ${factory}($$$ARGS)` } });
+		for (const match of matches) {
```
```suggestion
		for (const match of matches) {
```

NIT

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+
+	// 2️ Track variables assigned from factories
+	for (const factory of factoryBindings) {
+		const matches = rootNode.findAll({ rule: { pattern: `const $$$VAR = ${factory}($$$ARGS)` } });
```
Here it's only catch lexical declarations that have const. You may need a bigger query there.

Also add test for `let` & `var`

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+	const nodeRequires = getNodeRequireCalls(root, "node:zlib");
+	const nodeImports = getNodeImportStatements(root, "node:zlib");
+	const importNodes = [...nodeRequires, ...nodeImports];
+	if (!importNodes.length) return null;
```
```suggestion

	// If any import is found it's mean we can skip transformation on this file
	if (!importNodes.length) return null;
```

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,96 @@
+	const linesToRemove: Range[] = [];
+
+	// 1️ Find all zlib imports/requires
+	const nodeRequires = getNodeRequireCalls(root, "node:zlib");
+	const nodeImports = getNodeImportStatements(root, "node:zlib");
+	const importNodes = [...nodeRequires, ...nodeImports];
```
```suggestion
	const importNodes = [
		...getNodeRequireCalls(root, "node:zlib"),
		 ...getNodeImportStatements(root, "node:zlib")
	];
```

Since your aren't using theses variables somewhere else no needed to cache theme.
FYI: cache mean storing the result of a function in a variable to not having to recall the function.

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/workflow.yaml:
```diff
@@ -0,0 +1,25 @@
+    name: Apply AST Transformations
+    type: automatic
+    steps:
+      - name: Handle DEP0108 by replacing deprecated `zlib.bytesRead` with `zlib.bytesWritten` in Node.js transform streams`
```
valid

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/README.md:
```diff
@@ -0,0 +1,139 @@
+# `zlib.bytesRead` → `zlib.bytesWritten` DEP0108
+
+This codemod replace zlib.bytesRead with zlib.bytesWritten for consistent stream property naming. It's useful to migrate code that uses the deprecated property which has been removed.
+
+It replace zlib.bytesRead with zlib.bytesWritten in all zlib transform streams and it handle both CommonJS and ESM imports.
```
valid

AugustinMauroy commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,116 @@
+			
+			for (const match of matches) {
+				const varMatch = match.getMultipleMatches("VAR");
+				if (varMatch.length) {
```
```suggestion

				if (varMatch.length) {
```

AugustinMauroy reviewed (commented):
Wow impressive. That nice some small review

AugustinMauroy reviewed (approved):
WOW impressive first contribution. For me everything is covered.

I hope you enjoyed this contribution.

brunocroh commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,140 @@
+	const streamVariables: string[] = [];
+
+	for (const node of importNodes) {
+		const baseBind = resolveBindingPath(node, "$");
+
+		if (baseBind) {
+			for (const factory of ZLIB_FACTORIES) factoryBindings.push(`${baseBind}.${factory}`);
+		}
+		for (const factory of ZLIB_FACTORIES) {
+			const binding = resolveBindingPath(node, `$.${factory}`);
+			if (binding && !factoryBindings.includes(binding)) factoryBindings.push(binding);
+		}
```
```js
const baseBind = resolveBindingPath(node, "$");
```

I do not think the resolveBindingPath, can works in this way, you need to specify which binding you want to get the path, just '$' not means anything for this function.

So I think you code can be refactored to below one, if I missed something, let me know
```suggestion
		for (const factory of ZLIB_FACTORIES) {
			const binding = resolveBindingPath(node, `$.{factory}`);
			if (binding && !factoryBindings.includes(binding)) factoryBindings.push(binding);
		}
```

brunocroh commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	// 1.b Handle dynamic imports: `await import("node:zlib")`
+	const dynamicImports = rootNode.findAll({ rule: { pattern: "const $$$VAR = await import($$$MODULE)" } });
```
You just need to use `$$$`(Triple dollar sign) when more than on values will be found, in both case, just one will be existing so it needs to be written with `$`

```js
const dynamicImports = rootNode.findAll({ rule: { pattern: "const $VAR = await import($MODULE)" } });
```

btw, I think we have an utility that can resolve dynamic imports already, and this usage is not covering all cases, by example, if this line starts with `var` or `let` this not will match already.

Eliekhoury17 replied:
I did not find the utility that can resolve dynamic imports but I have covered the rest of the cases for `let` and `var` and replaced the `$$$` with `$` .

I committed the changes.

Eliekhoury17 replied:
I did not `resolve convesation`. If you know if the utility that can resolve dynamic imports exists and how can i import it  I can use it and modify my code, if not, what i commited work very well and i will `resolve convesation`.

brunocroh replied:
Yep I create a patch to demonstrade how to handle it, you can check it 

https://github.com/nodejs/userland-migrations/pull/224#pullrequestreview-3311187263

brunocroh commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,140 @@
+
+	// 2 Track variables assigned from factories (const, let, var)
+	for (const factory of factoryBindings) {
+		const patterns = [
+			`const $$$VAR = ${factory}($$$ARGS)`,
+			`let $$$VAR = ${factory}($$$ARGS)`,
+			`var $$$VAR = ${factory}($$$ARGS)`
+		];
```
You need to use binding that have been found for each "factory" because it can have a different name, and resolve binding path, will have the named used in the file.

Variable name can be preserved, I think we don't want to change user variable name at the end.

brunocroh commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,140 @@
+				const paramNames = param
+					.text()
+					.split(",")
+					.map((p) => p.replace(/\/\*.*\*\//, "").trim())
```
Could you explain the strategy behind this map?
What this regex is doing, and why you need a trim? I think the ast-grep already cover it, we don't need to make this clean-up on our side, did you find any special case that need it?

Eliekhoury17 replied:
I the end i used use `kind` to fetch functions and not a `pattern` which removed this part which was not needed anymore.

I committed the changes.

brunocroh commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,140 @@
+	// 4 Replace .bytesRead → .bytesWritten for function parameters
+	const funcPatterns = ["function $$$NAME($$$PARAMS) { $$$BODY }"];
+	for (const pattern of funcPatterns) {
+		const funcs = rootNode.findAll({ rule: { pattern } });
```
You need to use `kind` to fetch functions and not a `pattern`, it will not work in some scenarios.

Eliekhoury17 replied:
I used use `kind` to fetch functions and not a `pattern`

I committed the changes.

Eliekhoury17 commented:
Is it possible for you to explain a little bit what the link to the patch contains and what i need to change. I am new in theses things and I'm a little lost here on what to do in the code. @brunocroh

Eliekhoury17 commented:
I applied your patch on my code and all the test passes. So do I need to do something else than this?

brunocroh commented:
> you to explain a little bit what the link to the patch co

Oh, sorry, you just need to download the file on the link I share, and use `git apply name_of_file`, and git will apply the changes in the code for you.

brunocroh commented on recipes/zlib-bytesread-to-byteswritten/src/workflow.ts:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	// Step 4: Replace .bytesRead → .bytesWritten for function parameters
+	for (const kind of FUNC_KINDS) {
+		const funcs = rootNode.findAll({ rule: { kind } });
+
+		for (const func of funcs) {
+			const formalParamsNode = func
+				.children()
+				.find((child) => child.kind() === 'formal_parameters');
+
+			if (!formalParamsNode) continue;
+
+			const paramNames = findIdentifiers(formalParamsNode);
+
+			for (const paramName of paramNames) {
+				const matches = rootNode.findAll({
+					rule: { pattern: `${paramName}.bytesRead` },
+				});
+				for (const match of matches) {
+					edits.push(
+						match.replace(match.text().replace('.bytesRead', '.bytesWritten')),
+					);
+				}
+			}
+		}
+	}
```
I think the find Identifier util can re removed if we simplify the way the function parameters are fetched.

I created a patch for this one as well (and create a new testing scenario with a function that receive more than one parameter).

https://gist.githubusercontent.com/brunocroh/71cdfb6cc9ee4cfcd8e690ab9f888b3f/raw/25bb2e3b71e57f136ef7f81afda058d87986c4e2/gistfile0.txt

```suggestion
	for (const kind of FUNC_KINDS) {
		const funcs = rootNode.findAll({ rule: { kind } });

		for (const func of funcs) {
			const paramNames = func
				.field('parameters')
				?.findAll({ rule: { kind: 'identifier' } });

			for (const paramName of paramNames) {
				// replace member_expressions that use ${paramName}.bytesRead inside the function context
				const matches = func.findAll({
					rule: {
						kind: 'member_expression',
						pattern: `${paramName.text()}.bytesRead`,
					},
				});
				for (const match of matches) {
					edits.push(
						match.replace(match.text().replace('.bytesRead', '.bytesWritten')),
					);
				}
			}
		}
	}
```

Eliekhoury17 commented:
I checked everything, added the changes and modified one comment and committed the changes of your last patch and the tests and the code changes are good. For me everything looks good. @brunocroh

Eliekhoury17 commented:
Is there anything else you want me to do or modify in the code so it can be merged?

brunocroh commented:
> Is there anything else you want me to do or modify in the code before it can be merged?

I think all scenarios are covered. Great work, and thank you for your awesome contribution! 😁

Eliekhoury17 commented:
> I think all scenarios are covered. Great work, and thank you for your awesome contribution! 😁

Thank you all, @AugustinMauroy , @brunocroh and @JakobJingleheimer, for you help and feedback. I real enjoyed contributing to your open source project.
